README.md
drserv is an authenticated HTTP API for publishing .deb package to a repo

## API Usage

This service makes one endpoint available:

```
/v1/publish/{major.dist}/{minor.dist}/{component}/{filename.deb}
```

`major.dist` indicates the code name of the target Debian type distribution,
a string such as `squeeze` or `trusty`.

`minor.dist` is a somewhat special concept where different users of
a repository for a specific upstream Debian type distribution wants slightly
different versions of packages. Typical values of this part is `stable`,
`testing` or `unstable`.

`component` corresponds to the component concept in the upstream structure 
as documented in the [RepositoryFormat] and would be a string such as `main`
or `non-free`.

The content of the deb file to be published is to be provided as POST body
data.


[RepositoryFormat]: https://wiki.debian.org/RepositoryFormat


debian/changelog
drserv (0.1.3) unstable; urgency=low

  * debian/default
    - Add vars DRSERV_USER and DRSERV_GROUP
  * debian/upstart
    - Read user and group from /etc/default/drserv.
  * debian/postinst
    - Change the way how user is created.

 -- Client-Build <client-build@spotify.com>  Tue, 7 Jul 2015 12:12:01 +0000

drserv (0.1.2) unstable; urgency=low

  * drserv/server.py:
    - Give umask based permissions to received files.

 -- Client-Build <client-build@spotify.com>  Mon, 6 Jul 2015 10:54:15 +0000

drserv (0.1.1) unstable; urgency=low

  * debian/upstart:
    - Update umask to 002. Creating group writable directories is good for
      interacting with pre-drserv ways of uploading packages.

 -- Client-Build <client-build@spotify.com>  Fri, 29 May 2015 17:38:15 +0000

drserv (0.1.0) unstable; urgency=low

  * Initial release.

 -- Client-Build <client-build@spotify.com>  Fri, 29 May 2015 17:38:15 +0000


test/test_server.py
# coding=utf-8
# Copyright (c) 2015 Spotify AB
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import pytest
from drserv import server


@pytest.fixture
def cls():
    return server.DrservServer


class TestServerParsePath(object):
    @pytest.mark.randomize(parts=pytest.list_of(str, min_items=5, max_items=5),
                           min_length=3, ncalls=10)
    @pytest.mark.randomize(version=int, min_num=1, max_num=9999)
    def test_part_splitting(self, cls, parts, version):
        """
        Test splitting of paths with random string values.

        Because there are two randomize() decorators, the will be run the
        in an amount totalling (x + y) ^ 2 times. So, when the first one
        has ncalls=10 and the second has 1 they will run 100 times.

        """

        path = '/v1/publish/{0}/{1}/{2}/{3}-{4}_{5}.deb'.format(
            parts[0],
            parts[1],
            parts[2],
            parts[3],
            parts[4],
            version,
        )
        major, minor, component, filename = cls.parse_path(path)

        assert major == parts[0]
        assert minor == parts[1]
        assert component == parts[2]
        assert filename == '{0}-{1}_{2}.deb'.format(
            parts[3], parts[4], version
        )

    # TODO thiderman: All of these are raising the same exception. It's
    # probably okay, but if we want to know exactly what's going on we would
    # need to differentiate them more.
    def test_invalid_numeric_path(self, cls):
        with pytest.raises(ValueError):
            cls.parse_path(42)

    def test_invalid_unicode_path(self, cls):
        with pytest.raises(ValueError):
            cls.parse_path(u'€')

    def test_control_characters_disallowed(self, cls):
        with pytest.raises(ValueError):
            cls.parse_path('\x00')

    def test_relative_paths_disallowed(self, cls):
        with pytest.raises(ValueError):
            cls.parse_path('/v1/publish/../../../some-package_1.deb')

    def test_deb_file_suffix_is_required(self, cls):
        with pytest.raises(ValueError):
            cls.parse_path('/v1/publish/squeeze/stable/non-free/foo')

    def test_wrong_number_of_path_arguments(self, cls):
        with pytest.raises(server.HttpException):
            cls.parse_path('/v1/publish/squeeze/stable/foo.deb')


drserv/client.py
# Copyright (c) 2015 Spotify AB
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import argparse
import os
import sys
from crtauth import ssh
from crtauth import client
import requests

from six.moves.urllib import parse as urlparse


def _authenticate(base_url, username, private_key_filename):

    try:
        with open(private_key_filename) as f:
            signer = ssh.SingleKeySigner(f.read())
    except:
        sys.stderr.write(
            'ERROR: Key file must be a passphraseless private key '
            'generated by ssh-keygen')
        sys.exit(1)

    challenge = _auth_get(
        base_url, 'request:%s' % client.create_request(username))
    hostname = urlparse.urlparse(base_url).netloc
    if hostname.index(':') != -1:
        # netloc might contain port information as well
        hostname = hostname[:hostname.index(':')]
    response = client.create_response(challenge, hostname, signer)
    return _auth_get(base_url, 'response:' + response)


def _auth_get(base_url, value):
    response = requests.get("%s/_auth" % base_url,
                            headers={"X-CHAP": value})
    if not response.ok:
        raise Exception("Authentication request failed with status %d: %s"
                        % (response.status_code, response.text))
    return response.headers["X-CHAP"].split(":")[1]


def main():
    parser = argparse.ArgumentParser(
        'drserv-client',
        description='Deploy a package onto a drserv service'
    )
    parser.add_argument('--url', action='store', required=True,
                        help='the base url of the drserv service')

    parser.add_argument('--key-file', action='store', required=True,
                        help='the rsa private key used to authenticate')
    parser.add_argument('--auth-user', action='store', required=True,
                        help='the username to authenticate as')

    parser.add_argument('--major-dist', action='store', required=True)
    parser.add_argument('--minor-dist', action='store', required=True)
    parser.add_argument('--component', action='store', required=True)

    parser.add_argument('package_filename')

    args = parser.parse_args()
    token = _authenticate(args.url, args.auth_user, args.key_file)

    url = '%s/v1/publish/%s/%s/%s/%s' % (
        args.url,
        args.major_dist,
        args.minor_dist,
        args.component,
        os.path.basename(args.package_filename),
    )

    with open(args.package_filename) as f:
        response = requests.post(url, data=f,
                                 headers={'Authorization': 'chap:' + token})
        if response.ok:
            print('Upload succeeded')
        else:
            print('Fail: %d: %s' % (response.status_code, response.text))


if __name__ == '__main__':
    main()


drserv/async_server.py
# Copyright (c) 2015 Spotify AB
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import logging
import time
import collections
import subprocess
import yaml
import sys
import os

import asyncio
import blessings
import types
from aiohttp import web

from crtauth import server
from crtauth import key_provider


log = logging.getLogger(__name__)

PackageInfo = collections.namedtuple(
    'PackageInfo', 'major_dist minor_dist component file')

# size of the chunks to read from the socket to the HTTP client
BUFFER_SIZE = 8192


class AsyncServer:
    t = blessings.Terminal()

    def __init__(self, port, base_dir, index_command, auth_server):
        self.port = port
        self.base_dir = base_dir
        self.index_command = index_command
        self.auth_server = auth_server

        self.address = '127.0.0.1'

        if not os.path.exists(base_dir):
            os.makedirs(base_dir)

        # The regexp parsing takes care of input validation for the URIs
        uri = (
            '/v1/publish/'
            '{major:[a-z]+}/'
            '{minor:[a-z]+}/'
            '{component:[a-z-]+}/'
            r'{package:[^/_]+}_{version}.deb'
        )
        self.routes = (
            ('POST', uri, self.publish),
        )

    @asyncio.coroutine
    def setup_loop(self, loop):
        app = web.Application(loop=loop)

        for method, route, function in self.routes:
            app.router.add_route(
                method,
                route,
                self.endpoint(function, method, route),
            )

        log.info(
            "Server starting at {0}:{1}".format(self.address, self.port)
        )
        srv = yield from loop.create_server(
            app.make_handler(),
            self.address,
            self.port,
        )

        return srv

    def serve_forever(self):
        loop = asyncio.get_event_loop()
        setup_future = self.setup_loop(loop)
        loop.run_until_complete(setup_future)

        loop.run_forever()

    def endpoint(self, func, method, route):
        """
        Decorator method that takes care of calling and post processing
        responses.

        """

        def wrap(*args, **kwargs):
            uri = args[0].path
            log.debug(
                '{t.bold_black}>>{t.white} {method} {t.normal}{uri}'.format(
                    method=method,
                    uri=uri,
                    t=self.t
                )
            )

            body = func(*args, **kwargs)
            code = 200

            # POST requests will need to read from asyncio interfaces, and thus
            # their handler functions will need to `yield from` and return
            # generator objects. If this is the case, we need to yield from
            # them to get the actual body out of there.
            if isinstance(body, types.GeneratorType):  # pragma: nocover
                body = yield from body

            if isinstance(body, tuple):
                # If the result was a 2-tuple, use the second item as the
                # status code.
                body, code = body

            s = '{t.bold_black}<<{t.white} {method} {t.blue}{uri}{t.normal}: '
            if code >= 400:
                s += '{t.bold_red}'
            else:
                s += '{t.bold_green}'

            s += '{code}{t.normal}'
            log.info(
                s.format(
                    method=method,
                    uri=uri,
                    code=code,
                    t=self.t
                )
            )
            response = web.Response(
                body=body.encode(),
                status=code,
                headers={'content-type': 'text/plain'}
            )

            return response

        return asyncio.coroutine(wrap)

    def publish(self, request):
        match = request.match_info
        filename = '{package}_{version}.deb'.format(**match)

        log.debug(
            "Publishing {0} to {major}/{minor}/{component}".format(
                filename,
                **match
            )
        )

        # Construct the target directory
        target_dir = '{0}/{major}/pool/{minor}/{component}/{package}/'.format(
            self.base_dir,
            **match
        )
        os.makedirs(target_dir, exist_ok=True)

        # Grab the file data from the request. We do this before checking if
        # the path already exists since by nature of the HTTP protocol we need
        # to receive the entirety of the POST before we can send anything back.
        file_data = yield from request.post()

        target = os.path.join(target_dir, filename)
        if os.path.exists(target):
            return 'Target already exists: {0}'.format(target), 400

        with open(target, 'wb') as f:
            f.write(file_data['file'].file.read())

        # Run the indexer
        log.debug("Calling index command: %s" % self.index_command)
        subprocess.check_call(self.index_command)

        return "OK", 202


def setup_logging():
    logging.basicConfig(
        level=logging.DEBUG,
        format='%(asctime)s+0000 - %(name)s - %(levelname)s - %(message)s')
    logging.Formatter.converter = time.gmtime


def main():
    setup_logging()

    parser = argparse.ArgumentParser(
        'drserv-server',
        description='The drserv service'
    )
    parser.add_argument(
        '--config', action='store',
        default='/etc/drserv.yml',
        help='the config file'
    )

    config = read_config(parser.parse_args().config)

    auth_server = server.AuthServer(
        config['crtauth_secret'],
        key_provider.FileKeyProvider(config['keys_dir']),
        config['service_name'],
        lowest_supported_version=1
    )

    serv = AsyncServer(
        config['listen_port'],
        config['target_basedir'],
        config['index_command'],
        auth_server
    )

    serv.serve_forever()


def read_config(file_name):
    try:
        with open(file_name) as f:
            return yaml.load(f)
    except IOError as e:
        fail('Failed to open config {}: {}'.format(file_name, e.strerror))


def fail(message):
    print(message, file=sys.stderr)
    sys.exit(-1)


if __name__ == '__main__':
    main()


drserv/server.py
# Copyright (c) 2015 Spotify AB
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import print_function
import argparse
import hashlib
import logging
import urllib
from wsgiref import simple_server
import time
import collections
import subprocess
import random
import yaml
import six
import sys
import os
from crtauth import server
from crtauth import key_provider
from crtauth import wsgi


log = logging.getLogger(__name__)

PackageInfo = collections.namedtuple(
    'PackageInfo', 'major_dist minor_dist component file')

# size of the chunks to read from the socket to the HTTP client
BUFFER_SIZE = 8192


class HttpException(Exception):
    """
    Used to fail an incoming http request with a non-200 code and message
    """
    pass


class DrservServer(object):
    """
    DrservServer instances listens to a port, responds to API calls
    and
    """
    def __init__(self, port, base_dir, index_command, auth_server):
        log.info('Starting server listening to port %d', port)
        self.base_dir = base_dir
        if not os.path.exists(base_dir):
            os.makedirs(base_dir)
        self.index_command = index_command

        self.wsgi_server = simple_server.make_server(
            '', port, wsgi.CrtauthMiddleware(self.handle_request, auth_server))

    def serve_forever(self):
        self.wsgi_server.serve_forever()

    def handle_request(self, environ, start_response):
        try:
            path = urllib.unquote(environ['PATH_INFO'])
            if not path.startswith('/v1/publish'):
                raise HttpException('404 Not found')
            if environ['REQUEST_METHOD'] != 'POST':
                raise HttpException('405 Only POST allowed')
            pi = self.parse_path(path)
            log.debug("Publishing %s to %s/%s/%s" %
                      (pi.file, pi.major_dist, pi.minor_dist, pi.component))

            # PyTypeChecker is buggy in Intellij IDEA 14.1.1
            # noinspection PyTypeChecker
            target_dir = self.build_target_dir(self.base_dir, pi)
            if not os.path.exists(target_dir):
                os.makedirs(target_dir)
            target = os.path.join(target_dir, pi.file)
            temp_filename, checksum = self.store_post_data(
                int(environ['CONTENT_LENGTH']), environ['wsgi.input'],
                target)
            log.debug("Received data in file %s" % temp_filename)

            if os.path.exists(target):
                os.unlink(temp_filename)
                raise HttpException(
                    "400 refusing to overwrite existing file %s " % target)
            os.rename(temp_filename, target)
            log.debug("renamed %s into %s with sha256 %s"
                      % (temp_filename, target, checksum))
            log.debug("calling %s" % self.index_command)
            subprocess.check_call(self.index_command)
            log.debug("called %s" % self.index_command)

            start_response("202 Accepted", [('Content-type', 'text/plain')])
            return "OK\n",

        except HttpException as e:
            log.warning("Failed request: " + e.message)
            start_response(e.message,
                           [('Content-type', 'text/plain'),
                            ('Content-length', str(len(e.message) - 3))])
            return e.message[4:] + "\n",

    @staticmethod
    def build_target_dir(base_dir, package_info):
        file_parts = package_info.file.split("_")
        if len(file_parts) < 2 or not file_parts[0]:
            raise HttpException("400 Filename %s invalid, should be of format "
                                "NAME_VERSION.deb" % package_info.file)
        return os.path.join(
            base_dir, package_info.major_dist, 'pool', package_info.minor_dist,
            package_info.component, file_parts[0]
        )

    @staticmethod
    def store_post_data(length, to_read_from, final_destination):
        """
        Reads length bytes of POST data from to_read_from and store in a
        unique temporary file based on the final_destination name.
        """
        hasher = hashlib.sha256()
        # We roll our own temporary file here to get umask based permissions,
        # and we do it next to the actual destination file to honor any setuid
        # or setgid bits on its parent directory.
        temp_file_name = "{}.part-{}".format(final_destination,
                                             '.part-',
                                             random.randint(0, 100100100))
        with open(temp_file_name, 'w') as f:
            to_read = min(length, BUFFER_SIZE)
            while length:
                buf = to_read_from.read(to_read)
                length -= len(buf)
                to_read = min(length, BUFFER_SIZE)
                f.write(buf)
                hasher.update(buf)
        return f.name, hasher.hexdigest()

    @staticmethod
    def parse_path(path):
        """
        Sanity checking and parsing of the path provided with the API call
        """
        if not isinstance(path, six.string_types):
            raise ValueError('wrong type %s of parameter path' % type(path))
        for i, c in enumerate(path):
            if not 0x1f < ord(c) < 0x7f:
                raise ValueError('Char on pos %d not in valid range: "%s"'
                                 % (i, c))

        elements = [x for x in path.split('/') if x]
        for e in elements:
            if e == '..':
                raise ValueError('Invalid parameter: ' + e)
        if not elements[-1].endswith('.deb'):
            raise ValueError('Your filename needs to end with ".deb"')
        if len(elements) != 6:
            raise HttpException(
                '400 path needs to be of form /v1/publish/{major.dist}'
                '/{minor.dist}/{component}/{filename.deb}')
        return PackageInfo(elements[2], elements[3], elements[4], elements[5])


def setup_logging():
    logging.basicConfig(
        level=logging.DEBUG,
        format='%(asctime)s+0000 - %(name)s - %(levelname)s - %(message)s')
    logging.Formatter.converter = time.gmtime


def main():
    setup_logging()

    parser = argparse.ArgumentParser(
        'drserv-server',
        description='The drserv service'
    )
    parser.add_argument('--config', action='store',
                        default='/etc/drserv.yml',
                        help='the config file')

    config = read_config(parser.parse_args().config)

    auth_server = server.AuthServer(
        config['crtauth_secret'],
        key_provider.FileKeyProvider(config['keys_dir']),
        config['service_name'],
        lowest_supported_version=1
    )

    serv = DrservServer(
        config['listen_port'],
        config['target_basedir'],
        config['index_command'],
        auth_server
    )

    serv.serve_forever()


def read_config(file_name):
    try:
        with open(file_name) as f:
            return yaml.load(f)
    except IOError as e:
        fail('Failed to open config {}: {}'.format(file_name, e.strerror))


def fail(message):
    print(message, file=sys.stderr)
    sys.exit(-1)


if __name__ == '__main__':
    main()
